src/test/java/org/jenkinsci/plugins/workflow/cps/CpsScmFlowDefinitionTest.java

```diff
@@ -161,6 +167,44 @@
         r.assertLogContains("version one", b);
     }
 
+    @Ignore("Requires way to signal changeset to WorkflowRun")
+    @Test public void lightweightChangelog() throws Exception {
+        sampleRepo.init();
+        sampleRepo.write("flow.groovy", "echo 'version one'");
+        sampleRepo.git("add", "flow.groovy");
+        sampleRepo.git("commit", "--message=init");
+        WorkflowJob p = r.jenkins.createProject(WorkflowJob.class, "p");
+        GitStep step = new GitStep(sampleRepo.toString());
+        CpsScmFlowDefinition def = new CpsScmFlowDefinition(step.createSCM(), "flow.groovy");
+        def.setLightweight(true);
+        p.setDefinition(def);
+        // we add the SCMRevisionAction because we are hijacking a plain SCM as a proxy for a SCMSource backed SCM
+        WorkflowRun b = r.assertBuildStatusSuccess(p.scheduleBuild2(0,
+                new SCMRevisionAction(new AbstractGitSCMSource.SCMRevisionImpl(
+                        new SCMHead("master"), sampleRepo.head()
+                ))
+        ));
+        r.assertLogNotContains("Cloning the remote Git repository", b);
+        r.assertLogContains("Obtained flow.groovy from git " + sampleRepo, b);
+        r.assertLogContains("version one", b);
+        List<ChangeLogSet<? extends ChangeLogSet.Entry>> changeSets = b.getChangeSets();
+        assertEquals("First build traditionally has empty changes", Collections.emptyList(), changeSets);
+        sampleRepo.write("flow.groovy", "echo 'version two'");
+        sampleRepo.git("add", "flow.groovy");
+        sampleRepo.git("commit", "--message=update");
+        // we add the SCMRevisionAction because we are hijacking a plain SCM as a proxy for a SCMSource backed SCM
+        b = r.assertBuildStatusSuccess(p.scheduleBuild2(0,
+                new SCMRevisionAction(new AbstractGitSCMSource
+                .SCMRevisionImpl(new SCMHead("master"), sampleRepo.head()
+                ))
+        ));
+        changeSets = b.getChangeSets();
+        assertEquals(1, changeSets.size());
```

Comment:
So this is currently 0 not 1 because `WorkflowRun` is unaware of the additional `changelog0.xml`. What probably needs to happen is that there would be a higher level action that could be attached in a more generic fashion and then `WorkflowRun` would be aware of that action and use it in addition to `WorkflowRun.SCMCheckout` in order to build the changeSets

Reply:
Cannot `SCMRevisionAction` be used for this purpose? `WorkflowRun` would need to use `scm-api` and compare to the last build, instead of that being done here.